### PR TITLE
Fix #12770: do not resolve unversioned plugins to pre-releases

### LINKED ISSUE #12770: Unversioned plugin resolves to a pre-release (4.0.0-beta-1) and fails with NoSuchMethodError
### Summary

When a plugin has no version anywhere in the effective model, Maven 4 resolves it from the
repository's `RELEASE` metadata. In Maven's version semantics a `4.0.0-beta-N` **is** a release
(it is not a `SNAPSHOT`), so Maven 4 selects a pre-release plugin that was compiled against an
unstable Maven 4 API — and the build dies with a linkage error.

Maven 3 does not have this problem: it runs a plugin compatibility check, rejects the
incompatible version and falls back to the newest usable one.

### Reproducer

Minimal showcase with CI: https://github.com/aschemaven/maven4-unversioned-plugin-pick

Three files: a POM with no parent, no `<build>`, no `<pluginManagement>` and no plugin version
anywhere, one trivial class, and a workflow running the *same* command on three setups.

```bash
mvn source:jar        # nothing pinned anywhere
```

`maven-source-plugin` is not part of any lifecycle binding, so the version has to come from
metadata. Central currently publishes:

```xml
<latest>4.0.0-beta-1</latest>
<release>4.0.0-beta-1</release>
```

**Apache Maven 4.0.0-rc-6** (`6a8189b24518daa120539fa41ce12f2b48ec09a8`), JDK 17, empty local
repository:

```
[INFO] --- source:4.0.0-beta-1:jar (default-cli) @ unversioned-plugin-pick ---
[ERROR] Failed to execute goal org.apache.maven.plugins:maven-source-plugin:4.0.0-beta-1:jar
        (default-cli) on project unversioned-plugin-pick: Execution default-cli of goal
        org.apache.maven.plugins:maven-source-plugin:4.0.0-beta-1:jar failed:
        An API incompatibility was encountered while executing
        org.apache.maven.plugins:maven-source-plugin:4.0.0-beta-1:jar:
        java.lang.NoSuchMethodError: 'java.util.List
        org.apache.maven.api.services.ProjectManager.getCompileSourceRoots(
        org.apache.maven.api.Project, org.apache.maven.api.ProjectScope)'
```

**Apache Maven 3.9.16** and **3.10.0-rc-1**, same project, same command, same JDK:

```
[WARNING] Ignoring incompatible plugin version 4.0.0-beta-1:
          The plugin org.apache.maven.plugins:maven-source-plugin:4.0.0-beta-1
          has unmet prerequisites:
[INFO] Latest version of plugin org.apache.maven.plugins:maven-source-plugin failed compatibility check
[INFO] Looking for compatible RELEASE version of plugin org.apache.maven.plugins:maven-source-plugin
[INFO] Selected plugin org.apache.maven.plugins:maven-source-plugin:3.4.0
[INFO] BUILD SUCCESS
```

Maven 3 only escapes *this* case by accident. The beta declares Maven 4 prerequisites, so the
compatibility check rejects it. The version selection itself ignores pre-release qualifiers on
every line: with a plugin whose pre-release *is* compatible (staged repository, stable `1.0` and
pre-release `2.0-beta-1`, metadata naming `2.0-beta-1` as `<release>`), **3.9.16 and 4.0.0-rc-6
both select `2.0-beta-1`**. That is the staged scenario of the integration test below.

The CI workflow encodes exactly this contrast — `unversioned` green on 3.9.16 and 3.10.0-rc-1, **red** on 4.0.0-rc-6,
`pinned-workaround` green — and runs weekly, so it keeps reflecting whatever Central currently
calls `RELEASE`.

### Why this is worth fixing rather than waiting it out

The symptom disappears once a stable `maven-source-plugin` 4.x is released, but the mechanism
stays: any plugin whose newest published version is a pre-release becomes a trap for every
build that does not pin. Two changes made this more likely than it used to be:

1. The Maven 3 super POM pinned four commonly CLI-invoked plugins in `pluginManagement`
   (`maven-antrun-plugin`, `maven-assembly-plugin`, `maven-dependency-plugin`,
   `maven-release-plugin`). Maven 3.10 and 4.x dropped that block, so more plugins now fall
   through to `RELEASE` resolution.
2. The compatibility fallback that saves the Maven 3 run is not a guard against pre-releases —
   it only checks the plugin's declared prerequisites. Under Maven 4 the beta's prerequisites are
   formally satisfied (it *is* Maven 4), so nothing catches the API break and the user gets a raw
   `NoSuchMethodError` naming an internal API method, with no hint that a version selection
   caused it.

### Fix

Branches (fork, ready to become PRs — `https://github.com/aschemaven/maven`):

* `bugfix/unversioned-plugin-prerelease` — off `master` (4.1)
* `backport/40x-unversioned-plugin-prerelease` — off `maven-4.0.x`

`DefaultPluginVersionResolver` is byte-identical on both lines, so the change cherry-picks cleanly.

What it does: when a plugin has no declared version, prefer a version *without* a pre-release
qualifier whenever the repository offers one; fall back to pre-releases, then snapshots, so a
plugin that has only ever published pre-releases still resolves.

A pre-release is recognised through the version scheme itself rather than a list of qualifier
names: such a qualifier sorts *before* the version it qualifies (`1.0-beta-1 < 1.0`), while a
build or vendor qualifier does not (`1.0-jre > 1.0`).

Alternatives that were considered and are *not* in the patch, but may be worth having as well:

1. Report a comprehensible failure when an auto-selected plugin throws
   `NoSuchMethodError`/`NoClassDefFoundError` against `org.apache.maven.api.*` — name the version
   selection as the likely cause instead of surfacing the linkage error alone.
2. Re-introduce super POM `pluginManagement` pins for the handful of plugins commonly invoked from
   the CLI without a version (Maven 3.10 and 4.x dropped the block that Maven 3.9 still has).

### Integration test

`MavenITghUnversionedPluginPrereleaseTest` + `its/core-it-suite/src/test/resources/gh-unversioned-plugin-prerelease/`.

It does not depend on live Central metadata: the test stages its own plugin repository offering a
stable `1.0` and a pre-release `2.0-beta-1`, with `<release>2.0-beta-1</release>` in the metadata —
the same shape Central has today for `maven-source-plugin`. The project declares the plugin without
a version and the test asserts the marker file written by `1.0`.

Verified:

| distribution | result |
|---|---|
| `master` + fix | IT passes; full core IT suite green (1052 tests, 0 failures) |
| `master` without fix | IT fails: `Expected file was not found: target/touch-stable.txt` (the beta ran) |
| `maven-4.0.x` + fix | see the backport branch |

The IT class and resource directory are named after this issue once it has a number.

### PR DESCRIPTION
**Pull request by ascheman**
Fixes #12770.

When a plugin has no version anywhere in the effective model, `DefaultPluginVersionResolver`
takes the `RELEASE` version from repository metadata. In Maven's version semantics a
`4.0.0-beta-N` *is* a release, so any plugin whose newest published version is a pre-release
becomes a trap for every build that does not pin:

```
$ mvn source:jar                       # Apache Maven 4.0.0-rc-6, no version pinned anywhere
[INFO] --- source:4.0.0-beta-1:jar (default-cli) @ demo ---
[ERROR] Failed to execute goal org.apache.maven.plugins:maven-source-plugin:4.0.0-beta-1:jar …
        java.lang.NoSuchMethodError: 'java.util.List
        org.apache.maven.api.services.ProjectManager.getCompileSourceRoots(
        org.apache.maven.api.Project, org.apache.maven.api.ProjectScope)'
```

`maven-source-plugin` currently publishes `<release>4.0.0-beta-1</release>` on Central, and that
beta was built against a Maven 4 API that changed before the line stabilised.

The existing compatibility check does not catch this — the beta's declared prerequisites are
satisfied (it *is* Maven 4), it is only compiled against an API that moved. Maven 3 escapes this
particular case by accident, because there the same prerequisites fail; with a pre-release that
*is* compatible, 3.9.16 selects it just as readily.

#### Change

Prefer a stable version whenever the repository offers one; fall back to pre-releases, and then
to snapshots, only when it does not — so a plugin that has only ever published pre-releases still
resolves.

A pre-release is recognised through the version scheme itself rather than a list of qualifier
names: such a qualifier sorts *before* the version it qualifies (`1.0-beta-1 < 1.0`), while a
build or vendor qualifier does not (`1.0-jre > 1.0`).

#### Integration test

`MavenITghUnversionedPluginPrereleaseTest` stages its own plugin repository — stable `1.0`,
pre-release `2.0-beta-1`, metadata naming the pre-release as `<release>` — so it keeps testing the
behaviour after Central's `maven-source-plugin` metadata moves on.

| distribution | result |
|---|---|
| this branch | IT passes; full core IT suite green locally (1052 tests, 0 failures) |
| master without the fix | IT fails: `Expected file was not found: target/touch-stable.txt` |

The full workflow of this repository also passes on a fork of this branch — `full-build` and
`integration-tests` on ubuntu/macos/windows × JDK 17, 21 and 25, all green:
https://github.com/aschemaven/maven/actions/runs/32263655600

A standalone reproducer with public CI is at
https://github.com/aschemaven/maven4-unversioned-plugin-pick — the `4.0.0-rc-6` job is red, the
`3.9.16` and `3.10.0-rc-1` jobs are green, and a fourth job shows that pinning any concrete 3.x
version is the workaround.

#### Backport

A `maven-4.0.x` backport is prepared and will follow once this lands (`DefaultPluginVersionResolver`
is identical on both lines). It carries one extra commit adapting the IT to that line's older test
harness — a version-range constructor and a `TestSuiteOrdering` entry, neither of which exists on
master since #11251.

### CODE REVIEW

**Review by gnodet — COMMENTED** [review]
**APPROVE** ✅

Well-designed fix that prevents unversioned plugin declarations from resolving to pre-release versions when stable alternatives exist. The three-tier fallback (stable → pre-release → snapshot) is clean, and the approach of using the version scheme's own ordering to detect pre-releases (`version < baseVersion`) rather than maintaining a qualifier name list is robust and maintenance-free.

**Observations (non-blocking):**

1. The `isPreRelease` method uses `version.indexOf('-')` to locate the qualifier boundary, but Maven's `GenericVersionScheme` treats dots and dashes as equivalent separators (e.g., `1.0.beta.1` equals `1.0-beta-1`). A version published as `1.0.beta.1` (dot-separated, no dash) would not be detected as a pre-release. In practice this is inconsequential since Maven plugin versions on Central overwhelmingly use dash-separated qualifiers.

2. The PR description is exceptionally thorough — it clearly explains the problem, the approach, and includes a reproducible scenario with a standalone reproducer repository and CI evidence. This is a model contribution.

3. The decision to detect pre-releases via version scheme comparison rather than maintaining a list of qualifier names (alpha, beta, rc, etc.) is the right engineering call — it automatically adapts to the `GenericVersionScheme`'s well-known qualifier table.

_This review was generated by an AI agent (Claude Code) and may contain inaccuracies. Please verify all suggestions before applying._

_Claude Code on behalf of Guillaume Nodet_

**Comment by ascheman** [comment]
All five points are addressed in [`0e1f5c84f0`](https://github.com/apache/maven/pull/12771/commits/0e1f5c84f0581d0da24b4bd91f22fceae4b150a0). The four inline threads carry the same answers where they were raised.

**@gnodet**, on observation 1 of your [review](https://github.com/apache/maven/pull/12771#pullrequestreview-4974685145):

> The `isPreRelease` method uses `version.indexOf('-')` to locate the qualifier boundary, but Maven's
> `GenericVersionScheme` treats dots and dashes as equivalent separators (e.g., `1.0.beta.1` equals
> `1.0-beta-1`). A version published as `1.0.beta.1` (dot-separated, no dash) would not be detected as
> a pre-release.

That was a real gap, not just theory. `isPreRelease` located the
qualifier with `indexOf('-')`, so `1.0.beta.1` (and `1.0_alpha_1`) slipped through even though the
scheme considers them the same versions as `1.0-beta-1`. It now finds the boundary by scanning
tokens: everything up to the first non-numeric token is the base version, with `-`, `.` and `_`
treated alike. Both spellings are covered by the unit test.

**From the automated review:**

1. *Javadoc vs. snapshots* — kept the behaviour, fixed the description. `isPreRelease` deliberately
   answers "does this sort before its base version", which snapshots also satisfy; the callers
   classify snapshots first, so they never reach the pre-release bucket. The Javadoc now says
   exactly that instead of implying the method is limited to alpha/beta/milestone/rc.
2. *Inconsistent log kinds* — now `RELEASE` / `PRE-RELEASE` / `SNAPSHOT`.
3. *`hasStableVersion` and unparseable versions* — good catch: an unparseable entry counted as a
   stable fallback while the selection loop (which parses) could never pick it. It now requires the
   version to be parseable by the configured `VersionScheme`, so the "a stable one exists" decision
   matches what is actually selectable.
4. *No unit test for the selection order* — added five, driving `selectVersion` with a stubbed
   `MavenPluginManager` so the ordering is observed on its own: a stable version wins over the
   pre-release the metadata names, the newest stable is taken, a pre-release is used only when no
   stable exists, a snapshot is the last resort, and a stable `<release>` is taken as-is without a
   search.

The unit test class is now 29 cases and the IT is unchanged. Behaviourally the only difference to
the previous revision of this PR is that pre-releases written with `.` or `_` separators are now
recognised as such; every version that was selected before is still selected.

File: impl/maven-core/src/main/java/org/apache/maven/plugin/version/internal/DefaultPluginVersionResolver.java [hunk 1]
```diff
@@ -270,6 +269,61 @@ private void selectVersion(DefaultPluginVersionResult result, PluginVersionReque
...
+            return versionScheme
+                            .parseVersion(version)
+                            .compareTo(versionScheme.parseVersion(version.substring(0, qualifier)))
+                    < 0;
+        } catch (InvalidVersionSpecificationException e) {
+            return false;
+        }
+    }
+
     private boolean isCompatible(PluginVersionRequest request, String version) {
```
[the remark this answers is not in the document]
Comment (ascheman) [reply impl/maven-core/src/main/java/org/apache/maven/plugin/version/internal/DefaultPluginVersionResolver.java]: Option (b): the behaviour is deliberate, the Javadoc was not. `isPreRelease` answers "does this version sort before the version it qualifies", which snapshots satisfy too; every caller classifies snapshots first (`endsWith("-SNAPSHOT")`), so they never reach the pre-release bucket. Excluding them inside the predicate would duplicate that test and make the method's name disagree with its own ordering semantics. The Javadoc now says exactly this. https://github.com/apache/maven/pull/12771/commits/0e1f5c84f0581d0da24b4bd91f22fceae4b150a0

File: impl/maven-core/src/main/java/org/apache/maven/plugin/version/internal/DefaultPluginVersionResolver.java [hunk 2]
```diff
@@ -214,34 +232,15 @@ private void selectVersion(DefaultPluginVersionResult result, PluginVersionReque
...
-                    String ver = v.toString();
-                    if (isCompatible(request, ver)) {
-                        version = ver;
-                        repo = versions.versions.get(version);
-                        break;
-                    }
-                }
+            if (version == null) {
+                version = selectCompatible(request, snapshots, "SNAPSHOT");
+            }
```
[the remark this answers is not in the document]
Comment (ascheman) [reply impl/maven-core/src/main/java/org/apache/maven/plugin/version/internal/DefaultPluginVersionResolver.java:243]: Agreed — now `RELEASE` / `PRE-RELEASE` / `SNAPSHOT`. https://github.com/apache/maven/pull/12771/commits/0e1f5c84f0581d0da24b4bd91f22fceae4b150a0

File: impl/maven-core/src/main/java/org/apache/maven/plugin/version/internal/DefaultPluginVersionResolver.java [hunk 3]
```diff
@@ -270,6 +269,61 @@ private void selectVersion(DefaultPluginVersionResult result, PluginVersionReque
...
+    private boolean hasStableVersion(Versions versions) {
+        for (String ver : versions.versions.keySet()) {
+            if (!ver.endsWith("-SNAPSHOT") && !isPreRelease(ver)) {
+                return true;
+            }
+        }
+        return false;
+    }
+
+    /**
```
[the remark this answers is not in the document]
Comment (ascheman) [reply impl/maven-core/src/main/java/org/apache/maven/plugin/version/internal/DefaultPluginVersionResolver.java:308]: Good catch, that was a real inconsistency: an unparseable entry made `hasStableVersion` claim a stable fallback exists, while the selection loop (which parses) could never pick it. It now also requires the version to be parseable by the configured `VersionScheme`. https://github.com/apache/maven/pull/12771/commits/0e1f5c84f0581d0da24b4bd91f22fceae4b150a0

File: impl/maven-core/src/main/java/org/apache/maven/plugin/version/internal/DefaultPluginVersionResolver.java [hunk 4]
```diff
@@ -186,6 +186,21 @@ private void selectVersion(DefaultPluginVersionResult result, PluginVersionReque
...
+        if (version != null && isPreRelease(version) && hasStableVersion(versions)) {
+            logger.info(
+                    "Metadata of plugin {}:{} points at pre-release version {}, looking for a stable version",
+                    request.getGroupId(),
+                    request.getArtifactId(),
+                    version);
+            version = null;
+            repo = null;
+            searchPerformed = true;
+        }
```
[the remark this answers is not in the document]
Comment (ascheman) [reply impl/maven-core/src/main/java/org/apache/maven/plugin/version/internal/DefaultPluginVersionResolver.java:204]: Added five: a stable version wins over the pre-release the metadata names as `<release>`, the newest stable is taken, a pre-release is used only when no stable exists, a snapshot is the last resort, and a stable `<release>` is taken as-is without a search. They drive `selectVersion` with a stubbed `MavenPluginManager`, so the ordering is observed without the IT harness. https://github.com/apache/maven/pull/12771/commits/0e1f5c84f0581d0da24b4bd91f22fceae4b150a0

**Review by gnodet — APPROVED** [review]
The new commit cleanly addresses all review feedback: rewrites `isPreRelease` to handle dot/underscore separators (gnodet's observation), adds `isParseable` guard to `hasStableVersion`, fixes log label consistency, improves Javadoc, and adds five comprehensive unit tests for `selectVersion` ordering.

Notable points:
- `baseVersionOf` correctly handles all separator types (dash, dot, underscore) that `GenericVersionScheme` treats as equivalent, directly addressing the observation about dot-separated qualifiers like `1.0.beta.1`
- The `isParseable` guard in `hasStableVersion` fixes a real edge case where an unparseable version string could make `hasStableVersion` return true while the selection loop could never pick that version
- The five new `selectVersion` unit tests provide solid coverage of the three-tier selection order (stable → pre-release → snapshot)

_This review was generated by an AI agent (Claude Code) and may contain inaccuracies. Please verify all suggestions before applying._

_Claude Code on behalf of Guillaume Nodet_

**Comment by cstamas** [comment]
https://github.com/apache/maven-resolver/blob/master/maven-resolver-util/src/main/java/org/eclipse/aether/util/version/GenericQualifiers.java#L32

You are looking for positive qualifiers

**Comment by cstamas** [comment]
In fact, look at whole related commit, there are things you could reuse:
https://github.com/apache/maven-resolver/commit/283b542024f6a45d46c89cf0bcc5ec3495c7b8c6

**Comment by ascheman** [comment]
@cstamas — adopted, thank you, that is exactly the right hook.

> https://github.com/apache/maven-resolver/blob/master/maven-resolver-util/src/main/java/org/eclipse/aether/util/version/GenericQualifiers.java#L32
> You are looking for positive qualifiers

`GenericQualifiers` is public since maven-resolver 2.0.21, which this line already depends on, so no
version bump is needed. The whole hand-rolled base-version scan is gone:

```java
boolean isPreRelease(String version) {
    return GenericQualifiers.qualifier(version).orElse(GenericQualifiers.QUALIFIER_ZERO) < 0;
}
```

Checked against 2.0.21 before switching, and it is strictly better than what I had — it also handles
the short forms and does not match `rc` inside a word:

| version | qualifier |
|---|---|
| `1.0`, `3.15.0`, `33.7.0-jre`, `1.0-arc` | (none) |
| `1.0-ga`, `1.0-final` | 0 |
| `1.0-sp1` | +1 |
| `4.0.0-beta-1`, `1.0.beta.1`, `1.0-b2` | -4 |
| `1.0-M1`, `1.0-rc1`, `1.0-cr1` | -3 / -2 |
| `1.0-SNAPSHOT` | -1 |

Every existing test expectation held unchanged; I added `1.0-b2` and `1.0-arc` for the two cases the
reused implementation gets right by construction. https://github.com/apache/maven/pull/12771/commits/2d2ab470d7e351260e6732b72f6787d0521f1f99

I did not take `GenericQualifiersVersionFilter` from that commit — it filters a version *collection*
during collection, whereas this code path picks one version out of repository metadata for an
unversioned plugin. Happy to look again if you think it fits better there.

**Review by cstamas — APPROVED** [review]

**Comment by cstamas** [comment]
FYI there is also this snippet: https://github.com/maveniverse/toolbox/blob/main/shared/src/main/java/eu/maveniverse/maven/toolbox/shared/ArtifactVersionMatcher.java#L140-L163

Not in maven/resolver but in toolbox (so is not applicable here), just FYI

**Review by gnodet — APPROVED** [review]
Re-review after latest commit ("Reuse GenericQualifiers for the pre-release check").

Clean refactoring that correctly delegates `isPreRelease()` to `GenericQualifiers.qualifier()` from maven-resolver-util, replacing ~50 lines of hand-written base-version parsing and comparison logic with a direct query against the version scheme's own qualifier shift table. The behavioral equivalence holds for all realistic version strings.

The two new test inputs (`1.0-b2` for short-form beta, `1.0-arc` to confirm `rc` substring is not falsely matched) specifically validate GenericQualifiers boundary behavior and are good additions.

_This review was generated by an AI agent (Claude Code) and may contain inaccuracies. Please verify all suggestions before applying._

_Claude Code on behalf of Guillaume Nodet_
